[image_picker] fix bug, sometimes double click cancel button will crash

iheron opened the pull request:
## Description

Sometimes double click cancel button will crash.

## Related Issues

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

iheron commented:
@googlebot I fixed it.

cyanglaz commented on packages/image_picker/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 0.6.5
```
Bug fix usually only needs to bump the patch, so: 0.6.4+1
https://github.com/flutter/engine/pull/16935 for more info :)

iheron replied:
Fixed it

cyanglaz commented on packages/image_picker/ios/Classes/FLTImagePickerPlugin.m:
```diff
@@ -323,10 +323,12 @@ - (void)imagePickerController:(UIImagePickerController *)picker
-  self.result(nil);
+  if (self.result != nil) {
+    self.result(nil);
 
-  self.result = nil;
-  _arguments = nil;
+    self.result = nil;
+    _arguments = nil;
+  }
```
prefer escape early:
```
if (!self.result) {
  return;
}
self.result(nil);
self.result(nil);
_arguments = nil;
```

iheron replied:
Thanks for your advice.

iheron commented:
I don't know how to write test code to reproduce the bug. If you just open the page, multiple times and quickly click cancel button, can reproduce it.

cyanglaz commented:
@iheron We can have a XCTest to manually trigger 
`- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker` 
while self.result is nil.

iheron commented:
@cyanglaz I find a file ImagePickerPluginTests.m, it's not in the xcode project, I added it and added a function testPluginPickImageDeviceCancel. But I am not good at writing objective-c and XCTests. Could you please help me to complete the test? 😂

cyanglaz commented:
@iheron I would love to, but I have other priority works to do. I will eventually circle back to this but it will take a long time. :)

iheron commented:
Thanks.

zhongwuzw commented on packages/image_picker/ios/Tests/ImagePickerPluginTests.m:
```diff
@@ -61,6 +61,22 @@ - (void)testPluginPickVideoDeviceBack {
+  [plugin handleMethodCall:call
+                    result:^(id _Nullable r){
+                    }];
+  // todo test imagePickerControllerDidCancel when result is nil
```
I think we can just call two times of `imagePickerControllerDidCancel` here to simulate crash. @cyanglaz What do you think?

iheron replied:
I see #2688 fixed this, so need to close this PR?

zhongwuzw replied:
> I see #2688 fixed this, so need to close this PR?

Hey, you can rebase and add the test, we fix different crash. :)

iheron commented on packages/image_picker/ios/Tests/ImagePickerPluginTests.m:
```diff
@@ -66,6 +67,25 @@ - (void)testPluginPickVideoDeviceBack {
+  plugin.result = ^(id result) {
+
+  };
+  [plugin imagePickerControllerDidCancel:[plugin getImagePickerController]];
```
@zhongwuzw thanks for your help. I add `testPluginPickImageDeviceCancelClickMultipleTimes`, is this right?

cyanglaz replied:
it looks good. If you try to run the test without your change, it crashes, then you are good :)

iheron commented:
Is there anything that needs to be modified now? @cyanglaz
